contracts/basic-bandwidth-generation/contracts/BandwidthGenerator.sol

```diff
@@ -0,0 +1,91 @@
+pragma solidity 0.6.6;
+
+import "./CosmosToken.sol";
+import "./Gravity.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+/** 
+ * @title BandwidthGenerator
+ * @dev   Contract for generating Basic Bandwidth Credentials (BBCs) on the Nym cosmos blockchain, 
+ *        using ERC20 representations of NYM as payment. Utilises the Gravity Bridge for cross-chain payment. 
+ *        
+ *        Credentials represent a certain amount of bandwidth which can be sent through the Nym Mixnet. 
+ *        By default 1 NYM = 1 GB of bandwidth. The `MBPerToken` amount can be adjusted by the contract owner. 
+ * 
+ *        The amount of bandwidth bought is calculated according to the following formula: 
+ *        `(Token amount in 'wei' / 10**18) * MBPerToken`
+ */ 
+contract BandwidthGenerator is Ownable {
+    
+    CosmosERC20 public erc20;
+    Gravity     public gravityBridge; 
+    uint256     public MBPerToken; 
+    
+    event BBCredentialPurchased(
+        uint256 indexed Bandwidth,
+        uint256 indexed VerificationKey,
+        bytes   SignedVerificationKey, 
+        bytes32 indexed CosmosRecipient
```

Comment:
This field isn't expected by the client either, so it won't work with it in the current form. Initially, I considered that the person purchasing the bandwidth might not know at the time of buying it with what gateway it will use it, and so it would have some flexibility. In order to simplify the payment flow, we might consider that it does know all of this at the moment of purchase.
But even if it does, the client binary (the one calling to emit this event) is only aware of the gateway identity, not of the gateway's owner address, which is what would need to be put in this field. Some additional work would need to be done for the client to obtain that information (query the contract, filter the gateways, extract the owner)

Reply:
As per offline chat, going to leave this in and we can tweak the client